### ISSUE TITLE
Docker Hub images appear to install Torch with CUDA 11.7 instead of CUDA 11.8

### ISSUE THREAD

**Issue by philpax** [axolotl-ai-cloud/axolotl#444]
### Please check that this issue hasn't been reported before.

- [X] I searched previous [Bug Reports](https://github.com/OpenAccess-AI-Collective/axolotl/labels/bug) didn't find any similar reports.

### Expected Behavior

Using `docker pull winglian/axolotl:main-py3.10-cu118-2.0.1` or `docker pull winglian/axolotl:main-py3.9-cu118-2.0.1`, the Torch version installed should use CUDA 11.8, in accordance with the BNB install which is compiled for 11.8.

### Current behaviour

I'm using the Docker images with boxes from vast.ai. I've tested both 3.9 and 3.10, and both appear to be installed with Torch 11.7 by default:

```
root@C.6794639:~$ python3 -c 'import torch; print(torch.version.cuda)'
11.7

root@C.6794639:~$ python3 -m pip install --no-cache-dir -U torch==2.0.1+cu118 --extra-index-url https://download.pytorch.org/whl/cu118
# ...
Installing collected packages: torch
  Attempting uninstall: torch
    Found existing installation: torch 2.0.1
    Uninstalling torch-2.0.1:
      Successfully uninstalled torch-2.0.1
Successfully installed torch-2.0.1+cu118

root@C.6794639:~$ python3 -c 'import torch; print(torch.version.cuda)'
11.8
```

This means that Axolotl doesn't actually work for QLoRA training on these images out of the box, as BNB is compiled for CUDA 11.8 and gets confused by Torch being compiled for CUDA 11.7.

### Steps to reproduce

Pull down the image from the Docker Hub, then check the installed version of Torch with `python3 -c 'import torch; print(torch.version.cuda)'`.

### Possible solution

Unfortunately, Docker Hub doesn't give me much visibility into what's going on with the individual build steps, but my guess is that string interpolation is failing and Torch is being installed with 11.7. Installing Torch with 11.8 manually using the command above (noting that I took that from the base Dockerfile and manually interpolated the arguments) appears to work fine.

### Which Operating Systems are you using?

- [X] Linux
- [ ] macOS
- [ ] Windows

### Python Version

3.9 / 3.10

### axolotl branch-commit

main/9d629d8bff1c69a8d7753b18ab6029a7aaea2090

### Acknowledgements

- [X] My issue title is concise, descriptive, and in title casing.
- [X] I have searched the existing issues to make sure this bug has not been reported yet.
- [X] I am using the latest version of axolotl.
- [X] I have provided enough information for the maintainers to reproduce and diagnose the issue.

**Comment by NanoCode012** [comment]
Hello, thank you for the report. In the meantime, the docker build steps can be found here: https://github.com/OpenAccess-AI-Collective/axolotl/blob/main/docker/Dockerfile-base

**Comment by jphme** [comment]
> Hello, thank you for the report. In the meantime, the docker build steps can be found here: https://github.com/OpenAccess-AI-Collective/axolotl/blob/main/docker/Dockerfile-base

I had the same problem, unfortunately building from scratch leads to the same error.

Some dependency apparently installs the "wrong" pytorch version:

(in a container build completely from scratch incl. with the current commit):
```python
import torch
torch.__version__
'2.0.1+cu117'
```

Adding
```
ARG PYTORCH_VERSION="2.0.1"
RUN pip3 install --no-cache-dir -U torch==${PYTORCH_VERSION}+cu${CUDA} --extra-index-url https://download.pytorch.org/whl/cu$CUDA
```

in the "Dockerfile" after the axolotl installation fixes the problem, but there has to be some smarter way to "pin" the pytorch version?

**Comment by NanoCode012** [comment]
Weirdly, the base does install it https://github.com/OpenAccess-AI-Collective/axolotl/blob/35130711d634966de5aee6a94b3613551c521db6/docker/Dockerfile-base#L32

**Comment by tmm1** [comment]
The issue comes from the force-reinstall here:

https://github.com/OpenAccess-AI-Collective/axolotl/blob/35130711d634966de5aee6a94b3613551c521db6/docker/Dockerfile#L14C35-L14C35

**Comment by tmm1** [comment]
Maybe adding --no-dependencies to that line will help? Or just remove the --force-reinstall

**Comment by tmm1** [comment]
For example:

```
$ pip install --force-reinstall --no-dependencies numba
Successfully installed numba-0.57.1

$ pip install --force-reinstall numba
Successfully installed llvmlite-0.40.1 numba-0.57.1 numpy-1.24.4

**Comment by NanoCode012** [comment]
Yea, don't think we need `--force-reinstall`

**Comment by tmm1** [comment]
I ran into a weird situation today where my peft wasn't updating locally, and I ended up needing `--force-reinstall`.

I'll keep an eye on it.. and it seems like maybe what's actually required is `-U` according to https://stackoverflow.com/a/41012000/332798

**Comment by tmm1** [comment]
Okay I figured out the edge case.. basically if you install from git once, and you want to update to newer git, you _have_ to use `--force-reinstall`.

In the case of the dockerfile, since we start without any peft installed and install it via git once before axolotl, I think its fine.
